README.md
# Redshirts

Redshirts counts contributors in git repositories in the same way that Prisma Cloud Code Security counts them for developer-based pricing. You can use this tool to estimate the impact on credit consumption prior to connecting repos to the platform.

"Contributors" are users who commit code to your repos. The platform counts contributors, identified by email address, who have committed code to the default branch (including commits merged from other branches) to non-public repositories in the last 90 days. Users who contribute to multiple repos only get counted once.

Note that while this tool applies the same logic as the platform when identifying users, due to the timing of platform scans, differences in repo visibility for different access tokens, etc, these results may not exactly match those in the platform.

Report issues to your account team, PANW support, or at https://github.com/bridgecrewio/redshirts/issues

See [docs](./docs) for more information.

## Example

Suppose you have two repositories integrated in the platform, with the following commit history:

| Repo  | User              | Commit date |
|-------|-------------------|-------------|
| Repo1 | user1@example.com | 1 day ago   |
| Repo1 | user2@example.com | 2 days ago  |
| Repo1 | user1@example.com | 3 days ago  |
| Repo1 | user3@example.com | 99 days ago |
| Repo2 | user1@example.com | 1 day ago   |
| Repo2 | user4@example.com | 2 days ago  |

Repo1 has 2 unique contributors in the last 90 days (user1 and user2). Repo2 also has 2 contributors (user1 and user4). There are 3 total unique contributors to these repos in the last 90 days (user1, user2, user4).

# Contents

<!-- toc -->
* [Redshirts](#redshirts)
* [Contents](#contents)
* [Quickstart](#quickstart)
* [Installation and requirements](#installation-and-requirements)
* [Modes](#modes)
* [Examples](#examples)
* [Output](#output)
* [Support and troubleshooting](#support-and-troubleshooting)
* [Usage](#usage)
* [Commands](#commands)
<!-- tocstop -->

# Quickstart

1. Generate access token for your VCS
1. Install: `npm install -g @paloaltonetworks/redshirts`
1. Run: `redshirts github --token gph_xxx --orgs mygithuborg,myothergithuborg`

See [installation](./docs/installation.md) for more information.

# Installation and requirements

Install using `npm`: `npm install -g @paloaltonetworks/redshirts`

Requires [nodejs](https://nodejs.org/en/) v16 or higher. If you use nodejs for other purposes, we recommend using Node Version Manager ([*nix](https://github.com/nvm-sh/nvm), [windows](https://github.com/coreybutler/nvm-windows)).

See [installation](./docs/installation.md) for more information.

# Modes

Broadly, this tool supports two "modes" - one in which it connects directly to your version control system (VCS), and one which uses repositories that you have cloned locally.

In VCS mode, the tool will use your personal access token to connect to the VCS and obtain real-time commit data. This is the most accurate and comprehensive mode, and we recommend using it as a first choice. See [VCS mode](./docs/vcs-mode.md).

In local mode, the tool will scan directories you specify for cloned git repos, and use the `git log` command to count contributors. See [local mode](./docs/local-mode.md).

# Examples

`redshirts github --orgs mygithuborg --token GITHUB_PAT`

`redshirts bitbucket-server --host bitbucket.mycompany.com --projects ABC,XYZ --username myuser --token BB_TOKEN`

`redshirts gitlab --repo-file repos.txt --token GITLAB_PAT -o json`

`redshirts azuredevops --orgs myazureorg -o csv --sort contributors --exclude-empty --token ADO_PAT`

`redshirts local --dirs $HOME/repos,/tmp/another-repo --log-level debug`

See [Usage](#usage), [specifying repos](./docs/vcs-mode.md#specifying-repositories), and [specifying directories](./docs/local-mode.md#specifying-directories) for more examples.

# Output

There are three types of output:

* Console table (default)
* CSV
* JSON

For the console and CSV tabular output formats, you can use `--sort repo` (default) to sort the table by repo name alphabetically, or `--sort contributors` to sort by the contributor count, descending.

For all output formats, you can remove repos with no commits using the `--exclude-empty` option.

**Console table**

The console table shows the number of unique contributors per repository, and the total number of unique contributors. Note that the total will likely be less than the sum of the contributors per repo, because each contributor only gets counted once for all repos.

```
        Contributors per Repository
┌──────────────────────────┬──────────────┐
│ Repo                     │ Contributors │
├──────────────────────────┼──────────────┤
│ try-bridgecrew/cfngoat   │ 0            │
│ try-bridgecrew/codegoat  │ 1            │
│ try-bridgecrew/terragoat │ 2            │
└──────────────────────────┴──────────────┘
Total unique contributors: 2
```

**CSV**

This format is similar to the table above, but in CSV format. The total is included as part of the CSV table.

```
Repo,Unique contributors
try-bridgecrew/cfngoat,0
try-bridgecrew/codegoat,1
try-bridgecrew/terragoat,2
Total,2
```

**JSON**

The JSON format is suitable for programmatic processing and also includes more details than the tabular output. In particular, the output will include the list of all contributors and their last commit date, both per-repo and globally.

```
{
  "totalContributors": 2,
  "contributors": [
    {
      "email": "user1@example.com",
      "usernames": [
        "example-user1"
      ],
      "lastCommitDate": "2022-12-06T15:22:01Z"
    },
    {
      "email": "user2@example",
      "usernames": [
        "example-user2"
      ],
      "lastCommitDate": "2022-12-04T17:45:22Z"
    }
  ],
  "repos": {
    "try-bridgecrew/terragoat": {
      "totalContributors": 2,
      "contributors": [
        {
          "email": "user1@example.com",
          "usernames": [
            "mroberts-panw"
          ],
          "lastCommitDate": "2022-11-04T14:43:10Z"
        },
        {
          "email": "user2@example.com",
          "usernames": [
            "example-user2"
          ],
          "lastCommitDate": "2022-12-04T17:45:22Z"
        }
      ]
    },
    "try-bridgecrew/cfngoat": {
      "totalContributors": 0,
      "contributors": []
    },
    "try-bridgecrew/codegoat": {
      "totalContributors": 1,
      "contributors": [
        {
          "email": "user1@example.com",
          "usernames": [
            "example-user1",
          ],
          "lastCommitDate": "2022-12-06T15:22:01Z"
        }
      ]
    }
  }
}
```

# Support and troubleshooting

If you have issues, please first review the command output, which may contain specific error messages and suggestions.

Please report issues or enhancement requests to your account team, PANW support, or at https://github.com/bridgecrewio/redshirts/issues.

Include debug logs and command output with your submission. Debug logs will contain required troubleshooting information. Note that your token will not be included in the log output.

You can enable debug logs by using the `--log-level debug` argument or by setting the environment variable `LOG_LEVEL=debug`. Debug logs will be printed to the `stderr` stream, and normal output will be printed to the `stdout` stream.

# Usage

<!-- usage -->
```sh-session
$ npm install -g @paloaltonetworks/redshirts
$ redshirts COMMAND
running command...
$ redshirts (--version)
@paloaltonetworks/redshirts/0.6.0 darwin-x64 node-v16.13.1
$ redshirts --help [COMMAND]
USAGE
  $ redshirts COMMAND
...
```
<!-- usagestop -->

# Commands

<!-- commands -->
* [`redshirts azuredevops`](#redshirts-azuredevops)
* [`redshirts bitbucket`](#redshirts-bitbucket)
* [`redshirts bitbucket-server`](#redshirts-bitbucket-server)
* [`redshirts github`](#redshirts-github)
* [`redshirts github-server`](#redshirts-github-server)
* [`redshirts gitlab`](#redshirts-gitlab)
* [`redshirts gitlab-server`](#redshirts-gitlab-server)
* [`redshirts help [COMMAND]`](#redshirts-help-command)
* [`redshirts local`](#redshirts-local)

## `redshirts azuredevops`

Count active contributors for Azure DevOps repos

```
USAGE
  $ redshirts azuredevops -t <value> [--orgs <value>] [--projects <value>] [--skip-projects <value>] [-o
    summary|json|csv] [--ca-cert <value>] [--no-cert-verify] [--days <value>] [--sort repo|contributors]
    [--exclude-empty] [--include-public] [--log-level error|warn|info|debug|ERROR|WARN|INFO|DEBUG] [--repos <value>]
    [--repo-file <value>] [--skip-repos <value>] [--skip-repo-file <value>]

OUTPUT FLAGS
  -o, --output=(summary|json|csv)
      [default: summary] Output format for displaying data. Defaults to 'summary', which is a console-friendly tabular
      output. To see full committer details, instead of just the unique count (total and per repo), use 'json'.

  --exclude-empty
      Do not include repos with no commits in the output

  --log-level=(error|warn|info|debug|ERROR|WARN|INFO|DEBUG)
      [default: warn] Set the log level for the execution. Can also be set with the LOG_LEVEL environment variable. Use
      'debug' for granular logging, which will be required for any support cases. You can log individual responses using
      the environment variable REDSHIRTS_LOG_API_RESPONSES=true. You can also disable all logging by setting
      DISABLE_LOGS=true as an environment variable. This is not recommended, as you may miss important processing
      messages. All logs will be written to the stderr stream.

  --sort=(repo|contributors)
      [default: repo] The output field on which to sort for CSV or console output: alphabetically by repo fully qualified
      name, or by descending contributor count (ignored for JSON)

AUTHENTICATION FLAGS
  -t, --token=<value>  (required) An Azure DevOps user personal access token tied to the provided username. Can also be
                       supplied with the REDSHIRTS_TOKEN environment variable. This token must be tied to a user that
                       has sufficient visibility of the repo(s) being counted. See the description below for more
                       information about the token.

CONNECTION FLAGS
  --ca-cert=<value>  Path to certificate chain to use in HTTP requests. See
                     https://www.baeldung.com/linux/ssl-certificates for more information on obtaining a certificate
                     chain for your environment. Note that for some systems, the hostname for the API endpoint may be
                     different from the hostname you visit in a browser (e.g., api.github.com vs. github.com). Be sure
                     to obtain a certificate for the correct hostname.
  --no-cert-verify   Bypass SSL verification on HTTPS requests. WARNING: This is inherently dangerous and can expose
                     your data, including API tokens, in the event that your network path is compromised.

REPO SPECIFICATION FLAGS
  --days=<value>
      [default: 90] The number of days for which to fetch commit history. Defaults to 90, which is the value used in the
      Prisma Cloud platform. It is not recommended to change this except for experimentation purposes.

  --include-public
      The platform only counts contributors in private repos (and "internal" repos for some enterprise systems). If you
      wish to see contributors in public repos, for informational or other purposes, use this flag. This will also cause
      Redshirts to skip checking if a repo is public, so it can speed up the runtime if you know you are only supplying
      private repos, or mostly private repos, using --repos, --orgs, etc.

  --orgs=<value>
      Org names for which to fetch repos. Use the --repos and / or --projects options to add additional specific repos on
      top of those in the specified org(s). Use the --skip-repos and / or --skip-projects options to exclude individual
      repos or projects that are a part of these org(s).

  --projects=<value>
      Project names for which to fetch repos. Use the --repos option to add additional specific repos on top of those in
      the specified project(s). Use the --skip-repos option to exclude individual repos that are a part of these
      project(s).

  --repo-file=<value>
      The name of a file containing a list of repositories to fetch, one per line. See the --repos option for more
      details.

  --repos=<value>
      A comma-separated list of repository names to fetch, as fully qualified paths (e.g., owner/repo,
      group/subgroup/project, etc). For systems where the repo display name can differ from the repo URL / slug (e.g.,
      Gitlab), the repo slug must be used. Takes precedence over --repos-file. If no repo or org / group list is provided,
      then the script will run over all repos for the token provided. Note that for some systems, this is not equivalent
      to a list of all possible repos. See the notes for each VCS for more information. If used with --orgs, --groups,
      etc, then this is a list of *additional* repos to fetch beyond the repos associated with the specified org(s) /
      group(s).

  --skip-projects=<value>
      Project names for which to skip fetching repos. Use this option to skip projects that are part of the specified
      orgs. If the same project is included in both --projects and --skip-projects, it is skipped. If a repo from a
      skipped project is included in --repos, it is also still skipped.

  --skip-repo-file=<value>
      The name of a file containing repository names to skip, one per line. See --skip-repos for more details.

  --skip-repos=<value>
      A comma-separated list of repository names to skip - used in conjunction with the VCS --orgs, --groups, etc options
      to skip specific repos within those groupings, or to filter out repos if no repo / org list is specified at all. If
      the same repo is included in --repos and --skip-repos, then the repo will be skipped. Takes precedence over
      --skip-repo-file.

DESCRIPTION
  Count active contributors for Azure DevOps repos

  Note: you must provide --repos, --projects, and / or --orgs. Due to limitations in Azure DevOps APIs, it is not
  possible to use a personal access token to fetch all orgs and repos for a user.

  About authentication: you must use a personal access token, scoped to the appropriate organization(s), with the Code:
  Read scope. The token's user must have at least a Basic level of membership to the organizations being scanned. Note
  that this is higher than the default access level that may be provided when you add a user to a team. See:
  https://learn.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=
  azure-devops&tabs=Windows and https://learn.microsoft.com/en-us/azure/devops/organizations/accounts/add-organization-u
  sers?view=azure-devops&tabs=browser

  About rate limiting: For Azure DevOps, this tool will attempt to submit requests in a burst until a rate limit is hit,
  and then respect the rate limit reset information provided in the response. Azure DevOps does not consistently provide
  rate limit headers in the responses, and thus it is not possible to always avoid hitting a rate limit.

EXAMPLES
  $ redshirts azuredevops --token obnwxxx --repos org/project/repo,org/project/repo2,org/project2/repo

  $ redshirts azuredevops --token obnwxxx --orgs bridgecrewio,try-bridgecrew

  $ redshirts azuredevops --token obnwxxx --orgs bridgecrewio,try-bridgecrew --projects org/project
```

_See code: [dist/commands/azuredevops.ts](https://github.com/bridgecrewio/redshirts/blob/v0.6.0/dist/commands/azuredevops.ts)_

## `redshirts bitbucket`

Count active contributors for Bitbucket repos

```
USAGE
  $ redshirts bitbucket -u <value> -t <value> [--workspaces <value>] [-o summary|json|csv] [--ca-cert <value>]
    [--no-cert-verify] [--days <value>] [--sort repo|contributors] [--exclude-empty] [--include-public] [--log-level
    error|warn|info|debug|ERROR|WARN|INFO|DEBUG] [--repos <value>] [--repo-file <value>] [--skip-repos <value>]
    [--skip-repo-file <value>] [--requests-per-hour <value>]

OUTPUT FLAGS
  -o, --output=(summary|json|csv)
      [default: summary] Output format for displaying data. Defaults to 'summary', which is a console-friendly tabular
      output. To see full committer details, instead of just the unique count (total and per repo), use 'json'.

  --exclude-empty
      Do not include repos with no commits in the output

  --log-level=(error|warn|info|debug|ERROR|WARN|INFO|DEBUG)
      [default: warn] Set the log level for the execution. Can also be set with the LOG_LEVEL environment variable. Use
      'debug' for granular logging, which will be required for any support cases. You can log individual responses using
      the environment variable REDSHIRTS_LOG_API_RESPONSES=true. You can also disable all logging by setting
      DISABLE_LOGS=true as an environment variable. This is not recommended, as you may miss important processing
      messages. All logs will be written to the stderr stream.

  --sort=(repo|contributors)
      [default: repo] The output field on which to sort for CSV or console output: alphabetically by repo fully qualified
      name, or by descending contributor count (ignored for JSON)

AUTHENTICATION FLAGS
  -t, --token=<value>     (required) A Bitbucket access token tied to the provided username. Can also be supplied with
                          the REDSHIRTS_TOKEN environment variable. This token must be tied to a user that has
                          sufficient visibility of the repo(s) being counted. See the description below for how to
                          create this token.
  -u, --username=<value>  (required) Your Bitbucket username associated with the provided app token

CONNECTION FLAGS
  --ca-cert=<value>
      Path to certificate chain to use in HTTP requests. See https://www.baeldung.com/linux/ssl-certificates for more
      information on obtaining a certificate chain for your environment. Note that for some systems, the hostname for the
      API endpoint may be different from the hostname you visit in a browser (e.g., api.github.com vs. github.com). Be
      sure to obtain a certificate for the correct hostname.

  --no-cert-verify
      Bypass SSL verification on HTTPS requests. WARNING: This is inherently dangerous and can expose your data, including
      API tokens, in the event that your network path is compromised.

  --requests-per-hour=<value>
      [default: 1000] The maximum number of requests to the server per hour. Requests will be throttled and spread to
      achieve this limit. For example, if you specify a value of 3600 here, approximately one request per second will be
      submitted. Use this to adjust the throttling of the API calls in the event that the rate limit is being consumed by
      other sources simultaneously. If you are running in a self-hosted server environment without API rate limits, you
      can also set this to a very high number to effectively disable throttling, but this may impact server performance.

REPO SPECIFICATION FLAGS
  --days=<value>
      [default: 90] The number of days for which to fetch commit history. Defaults to 90, which is the value used in the
      Prisma Cloud platform. It is not recommended to change this except for experimentation purposes.

  --include-public
      The platform only counts contributors in private repos (and "internal" repos for some enterprise systems). If you
      wish to see contributors in public repos, for informational or other purposes, use this flag. This will also cause
      Redshirts to skip checking if a repo is public, so it can speed up the runtime if you know you are only supplying
      private repos, or mostly private repos, using --repos, --orgs, etc.

  --repo-file=<value>
      The name of a file containing a list of repositories to fetch, one per line. See the --repos option for more
      details.

  --repos=<value>
      A comma-separated list of repository names to fetch, as fully qualified paths (e.g., owner/repo,
      group/subgroup/project, etc). For systems where the repo display name can differ from the repo URL / slug (e.g.,
      Gitlab), the repo slug must be used. Takes precedence over --repos-file. If no repo or org / group list is provided,
      then the script will run over all repos for the token provided. Note that for some systems, this is not equivalent
      to a list of all possible repos. See the notes for each VCS for more information. If used with --orgs, --groups,
      etc, then this is a list of *additional* repos to fetch beyond the repos associated with the specified org(s) /
      group(s).

  --skip-repo-file=<value>
      The name of a file containing repository names to skip, one per line. See --skip-repos for more details.

  --skip-repos=<value>
      A comma-separated list of repository names to skip - used in conjunction with the VCS --orgs, --groups, etc options
      to skip specific repos within those groupings, or to filter out repos if no repo / org list is specified at all. If
      the same repo is included in --repos and --skip-repos, then the repo will be skipped. Takes precedence over
      --skip-repo-file.

  --workspaces=<value>
      A comma separated list of workspace and / or usernames for which to fetch repos. Use the --repos option to add
      additional specific repos on top of those in the specified workspace(s). Use the --skip-repos option to exclude
      individual repos that are a part of these workspace(s).

DESCRIPTION
  Count active contributors for Bitbucket repos

  About authentication: this tool uses an app password with the following scopes:

  - Repositories: read
  - Account: read (not required if you always use --repos and / or --workspaces)

  See: https://support.atlassian.com/bitbucket-cloud/docs/app-passwords/


CHANGELOG.md
# Changelog

## [0.6.0](https://github.com/bridgecrewio/redshirts/compare/redshirts-v0.5.1...redshirts-v0.6.0) (2023-06-06)


### Features

* filter out all github noreply email addresses ([#81](https://github.com/bridgecrewio/redshirts/issues/81)) ([f847416](https://github.com/bridgecrewio/redshirts/commit/f8474167b6233ef5ee639facb066de036e52dc80))

## [0.5.1](https://github.com/bridgecrewio/redshirts/compare/redshirts-v0.5.0...redshirts-v0.5.1) (2023-05-23)


### Features

* Add bitbucket server bearer token auth and no-cert-verify flag ([#78](https://github.com/bridgecrewio/redshirts/issues/78)) ([40728c9](https://github.com/bridgecrewio/redshirts/commit/40728c9db4cd0aceca26431246f9d88224280dca))

## [0.5.0](https://github.com/bridgecrewio/redshirts/compare/redshirts-v0.4.1...redshirts-v0.5.0) (2023-05-17)


### Features

* add 'did you mean' suggestion plugin ([4efb494](https://github.com/bridgecrewio/redshirts/commit/4efb494a129ef749f9346cacb479435112baba88))
* add environment variable for supplying token ([#29](https://github.com/bridgecrewio/redshirts/issues/29)) ([6ce4118](https://github.com/bridgecrewio/redshirts/commit/6ce411827a6ea49d739518c8e22771b878d6fb94))
* add update checker plugin ([63cb018](https://github.com/bridgecrewio/redshirts/commit/63cb018ee6a94fc472db40c356f51776e6307329))
* Cleanup and rationalization ([#23](https://github.com/bridgecrewio/redshirts/issues/23)) ([c0eb353](https://github.com/bridgecrewio/redshirts/commit/c0eb35338033600262626351b5cc8d308075da90))
* log more details from failed API calls ([#30](https://github.com/bridgecrewio/redshirts/issues/30)) ([6bf92a6](https://github.com/bridgecrewio/redshirts/commit/6bf92a63e6eed53fd274c155837672f22005676f))
* only check the default branch for bitbucket and git log ([#19](https://github.com/bridgecrewio/redshirts/issues/19)) ([7b3dba0](https://github.com/bridgecrewio/redshirts/commit/7b3dba0ce9cdd9d1b1938eae4efe2840f4c09f52))
* Release 0.4.0 ([#73](https://github.com/bridgecrewio/redshirts/issues/73)) ([33d5cc3](https://github.com/bridgecrewio/redshirts/commit/33d5cc34c540f46bc172ec8f24b13bdbd44342a8))
* Update bot exclusion list ([#45](https://github.com/bridgecrewio/redshirts/issues/45)) ([7494641](https://github.com/bridgecrewio/redshirts/commit/7494641a54c7f63e9f1f6dd9700498ffcc8782fa))


### Bug Fixes

* actions runner ([b49a186](https://github.com/bridgecrewio/redshirts/commit/b49a186f42b7001aa5dd3ca87d36cab9a72cd48a))
* add additional environment detail to debug logs ([#75](https://github.com/bridgecrewio/redshirts/issues/75)) ([8c52147](https://github.com/bridgecrewio/redshirts/commit/8c52147a531e0c0a53e79f33d90395bc915da538))
* add registry url ([7ded447](https://github.com/bridgecrewio/redshirts/commit/7ded447782a49e659545acbdd34a4ef797c9b00d))
* change package version ([ea18a93](https://github.com/bridgecrewio/redshirts/commit/ea18a93c67a3eeb88b7e71b5d2c3ce5a6255528b))
* fix env var name for logging API responses ([#70](https://github.com/bridgecrewio/redshirts/issues/70)) ([907565c](https://github.com/bridgecrewio/redshirts/commit/907565cad1a1ce95f84786262b4c104b8f1aee12))
* Get and use default branch for Azure ([#71](https://github.com/bridgecrewio/redshirts/issues/71)) ([f1a4cd3](https://github.com/bridgecrewio/redshirts/commit/f1a4cd3495c417058ff18138f41a0488b2784d60))
* ignore duplicate explicit repos ([1d6c0d6](https://github.com/bridgecrewio/redshirts/commit/1d6c0d6c9338a2bece0de79cc12b43c0a86ffda1))
* minor type and flag fixes ([#40](https://github.com/bridgecrewio/redshirts/issues/40)) ([103e73c](https://github.com/bridgecrewio/redshirts/commit/103e73c424cce3f1496f8aa4a763f1c0eb9a61fd))
* Use author, not committer, for BB server and gitlab ([#36](https://github.com/bridgecrewio/redshirts/issues/36)) ([1dfd6aa](https://github.com/bridgecrewio/redshirts/commit/1dfd6aa3f746b417da2efc50b48bf31372952f32))
* use self-hosted runner ([bd3b99b](https://github.com/bridgecrewio/redshirts/commit/bd3b99b6171faf5d6d626b308635357f3508f691))
* use windows local path separator ([#27](https://github.com/bridgecrewio/redshirts/issues/27)) ([fd9fdc3](https://github.com/bridgecrewio/redshirts/commit/fd9fdc3b6458d89804f96c53a30292370a34c9e6))

## [0.4.0](https://github.com/bridgecrewio/redshirts/compare/redshirts-v0.3.1...redshirts-v0.4.0) (2023-05-16)


### Features

* Release 0.4.0 ([#73](https://github.com/bridgecrewio/redshirts/issues/73)) ([33d5cc3](https://github.com/bridgecrewio/redshirts/commit/33d5cc34c540f46bc172ec8f24b13bdbd44342a8))

## [0.3.1](https://github.com/bridgecrewio/redshirts/compare/redshirts-v0.3.0...redshirts-v0.3.1) (2023-05-16)


### Bug Fixes

* fix env var name for logging API responses ([#70](https://github.com/bridgecrewio/redshirts/issues/70)) ([907565c](https://github.com/bridgecrewio/redshirts/commit/907565cad1a1ce95f84786262b4c104b8f1aee12))
* Get and use default branch for Azure ([#71](https://github.com/bridgecrewio/redshirts/issues/71)) ([f1a4cd3](https://github.com/bridgecrewio/redshirts/commit/f1a4cd3495c417058ff18138f41a0488b2784d60))

## [0.3.0](https://github.com/bridgecrewio/redshirts/compare/redshirts-v0.2.0...redshirts-v0.3.0) (2023-01-12)


### Features

* add 'did you mean' suggestion plugin ([4efb494](https://github.com/bridgecrewio/redshirts/commit/4efb494a129ef749f9346cacb479435112baba88))
* add environment variable for supplying token ([#29](https://github.com/bridgecrewio/redshirts/issues/29)) ([6ce4118](https://github.com/bridgecrewio/redshirts/commit/6ce411827a6ea49d739518c8e22771b878d6fb94))
* add update checker plugin ([63cb018](https://github.com/bridgecrewio/redshirts/commit/63cb018ee6a94fc472db40c356f51776e6307329))
* log more details from failed API calls ([#30](https://github.com/bridgecrewio/redshirts/issues/30)) ([6bf92a6](https://github.com/bridgecrewio/redshirts/commit/6bf92a63e6eed53fd274c155837672f22005676f))


### Bug Fixes


src/common/flags.ts
import { Flags } from '@oclif/core';
import { OptionFlag } from '@oclif/core/lib/interfaces';
import { LOG_API_RESPONSES_ENV } from './api-manager';
import { HelpGroup, OutputFormat, Protocol, SortField } from './types';
import { DEFAULT_DAYS, DEFAULT_LOG_LEVEL, DISABLE_LOG_ENV_VAR, LOG_LEVELS } from './utils';

export const commonFlags = {
    output: Flags.enum({
        description:
            "Output format for displaying data. Defaults to 'summary', which is a console-friendly tabular output. To see full committer details, instead of just the unique count (total and per repo), use 'json'.",
        char: 'o',
        options: Object.values(OutputFormat),
        default: OutputFormat.Summary,
        helpGroup: HelpGroup.OUTPUT,
    }),
    'ca-cert': Flags.file({
        description:
            'Path to certificate chain to use in HTTP requests. See https://www.baeldung.com/linux/ssl-certificates for more information on obtaining a certificate chain for your environment. Note that for some systems, the hostname for the API endpoint may be different from the hostname you visit in a browser (e.g., api.github.com vs. github.com). Be sure to obtain a certificate for the correct hostname.',
        required: false,
        helpGroup: HelpGroup.CONNECTION,
    }),
    'no-cert-verify': Flags.boolean({
        description:
            'Bypass SSL verification on HTTPS requests. WARNING: This is inherently dangerous and can expose your data, including API tokens, in the event that your network path is compromised.',
        helpGroup: HelpGroup.CONNECTION,
    }),
    days: Flags.integer({
        description:
            'The number of days for which to fetch commit history. Defaults to 90, which is the value used in the Prisma Cloud platform. It is not recommended to change this except for experimentation purposes.',
        required: false,
        default: DEFAULT_DAYS,
        helpGroup: HelpGroup.REPO_SPEC,
    }),
    sort: Flags.enum({
        description:
            'The output field on which to sort for CSV or console output: alphabetically by repo fully qualified name, or by descending contributor count (ignored for JSON)',
        options: Object.values(SortField),
        default: SortField.REPO,
        helpGroup: HelpGroup.OUTPUT,
    }),
    'exclude-empty': Flags.boolean({
        description: 'Do not include repos with no commits in the output',
        helpGroup: HelpGroup.OUTPUT,
    }),
    'include-public': Flags.boolean({
        description:
            'The platform only counts contributors in private repos (and "internal" repos for some enterprise systems). If you wish to see contributors in public repos, for informational or other purposes, use this flag. This will also cause Redshirts to skip checking if a repo is public, so it can speed up the runtime if you know you are only supplying private repos, or mostly private repos, using --repos, --orgs, etc.',
        helpGroup: HelpGroup.REPO_SPEC,
    }),
    'log-level': Flags.enum({
        description: `Set the log level for the execution. Can also be set with the LOG_LEVEL environment variable. Use 'debug' for granular logging, which will be required for any support cases. You can log individual responses using the environment variable ${LOG_API_RESPONSES_ENV}=true. You can also disable all logging by setting ${DISABLE_LOG_ENV_VAR}=true as an environment variable. This is not recommended, as you may miss important processing messages. All logs will be written to the stderr stream.`,
        options: [...LOG_LEVELS, ...LOG_LEVELS.map((l: string) => l.toUpperCase())],
        env: 'LOG_LEVEL',
        default: DEFAULT_LOG_LEVEL,
        helpGroup: HelpGroup.OUTPUT,
    }),
};

// these are common to all the VCSes but not to local, so keep them separate
export const repoFlags = {
    repos: Flags.string({
        description:
            'A comma-separated list of repository names to fetch, as fully qualified paths (e.g., owner/repo, group/subgroup/project, etc). For systems where the repo display name can differ from the repo URL / slug (e.g., Gitlab), the repo slug must be used. Takes precedence over --repos-file. If no repo or org / group list is provided, then the script will run over all repos for the token provided. Note that for some systems, this is not equivalent to a list of all possible repos. See the notes for each VCS for more information. If used with --orgs, --groups, etc, then this is a list of *additional* repos to fetch beyond the repos associated with the specified org(s) / group(s).',
        required: false,
        helpGroup: HelpGroup.REPO_SPEC,
    }),
    'repo-file': Flags.file({
        description:
            'The name of a file containing a list of repositories to fetch, one per line. See the --repos option for more details.',
        required: false,
        helpGroup: HelpGroup.REPO_SPEC,
    }),
    'skip-repos': Flags.string({
        description:
            'A comma-separated list of repository names to skip - used in conjunction with the VCS --orgs, --groups, etc options to skip specific repos within those groupings, or to filter out repos if no repo / org list is specified at all. If the same repo is included in --repos and --skip-repos, then the repo will be skipped. Takes precedence over --skip-repo-file.',
        required: false,
        helpGroup: HelpGroup.REPO_SPEC,
    }),
    'skip-repo-file': Flags.file({
        description:
            'The name of a file containing repository names to skip, one per line. See --skip-repos for more details.',
        required: false,
        helpGroup: HelpGroup.REPO_SPEC,
    }),
};

// ... but combine them for convenience
export const vcsFlags = {
    ...commonFlags,
    ...repoFlags,
};

export const vcsServerFlags = {
    hostname: Flags.string({
        description:
            'The hostname of your server, e.g. `git.mycompany.com`. Do not include the port and protocol here (see --port and --protocol).',
        char: 'h',
        required: true,
        helpGroup: HelpGroup.CONNECTION,
    }),
    port: Flags.integer({
        description: 'The port of your server, if not a standard port (443 for https, or 80 for http).',
        char: 'p',
        required: false,
        helpGroup: HelpGroup.CONNECTION,
    }),
    protocol: Flags.enum({
        description:
            'Protocol for your server, https (default) or http. Affects the default port value if you do not specify a port. You must specify a protocol here, not in the hostname.',
        options: Object.values(Protocol),
        required: false,
        default: Protocol.HTTPS,
        helpGroup: HelpGroup.CONNECTION,
    }),
};

export const throttlingFlags = {
    'requests-per-hour': Flags.integer({
        description:
            'The maximum number of requests to the server per hour. Requests will be throttled and spread to achieve this limit. For example, if you specify a value of 60 here, one request per minute will be submitted. Use this to adjust the throttling of the API calls in the event that the rate limit is being consumed by other sources simultaneously.',
        required: false,
        default: 1000,
        helpGroup: HelpGroup.CONNECTION,
    }),
};

export const getThrottlingFlag = (defaultRequestsPerHour: number): { 'requests-per-hour': OptionFlag<number> } => {
    return {
        'requests-per-hour': Flags.integer({
            description:
                'The maximum number of requests to the server per hour. Requests will be throttled and spread to achieve this limit. For example, if you specify a value of 3600 here, approximately one request per second will be submitted. Use this to adjust the throttling of the API calls in the event that the rate limit is being consumed by other sources simultaneously. If you are running in a self-hosted server environment without API rate limits, you can also set this to a very high number to effectively disable throttling, but this may impact server performance.',
            required: false,
            default: defaultRequestsPerHour,
            helpGroup: HelpGroup.CONNECTION,
        }),
    };
};
